// @duckdb/node-api/lib/DuckDBDataChunk.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DuckDBDataChunk = void 0;
const node_bindings_1 = __importDefault(require("@duckdb/node-bindings"));
const DuckDBVector_1 = require("./DuckDBVector");
class DuckDBDataChunk {
    chunk;
    vectors = [];
    constructor(chunk) {
        this.chunk = chunk;
    }
    static create(types, rowCount) {
        const chunk = new DuckDBDataChunk(node_bindings_1.default.create_data_chunk(types.map((t) => t.toLogicalType().logical_type)));
        if (rowCount != undefined) {
            chunk.rowCount = rowCount;
        }
        return chunk;
    }
    reset() {
        node_bindings_1.default.data_chunk_reset(this.chunk);
    }
    get columnCount() {
        return node_bindings_1.default.data_chunk_get_column_count(this.chunk);
    }
    get rowCount() {
        return node_bindings_1.default.data_chunk_get_size(this.chunk);
    }
    set rowCount(count) {
        node_bindings_1.default.data_chunk_set_size(this.chunk, count);
    }
    getColumnVector(columnIndex) {
        if (this.vectors[columnIndex]) {
            return this.vectors[columnIndex];
        }
        const vector = DuckDBVector_1.DuckDBVector.create(node_bindings_1.default.data_chunk_get_vector(this.chunk, columnIndex), this.rowCount);
        this.vectors[columnIndex] = vector;
        return vector;
    }
    visitColumnValues(columnIndex, visitValue) {
        const vector = this.getColumnVector(columnIndex);
        const type = vector.type;
        for (let rowIndex = 0; rowIndex < vector.itemCount; rowIndex++) {
            visitValue(vector.getItem(rowIndex), rowIndex, columnIndex, type);
        }
    }
    getColumnValues(columnIndex) {
        const values = [];
        this.visitColumnValues(columnIndex, (value) => values.push(value));
        return values;
    }
    convertColumnValues(columnIndex, converter) {
        const convertedValues = [];
        const type = this.getColumnVector(columnIndex).type;
        this.visitColumnValues(columnIndex, (value) => convertedValues.push(converter.convertValue(value, type)));
        return convertedValues;
    }
    setColumnValues(columnIndex, values) {
        const vector = this.getColumnVector(columnIndex);
        if (vector.itemCount !== values.length) {
            throw new Error(`number of values must equal chunk row count`);
        }
        for (let i = 0; i < values.length; i++) {
            vector.setItem(i, values[i]);
        }
        vector.flush();
    }
    visitColumns(visitColumn) {
        const columnCount = this.columnCount;
        for (let columnIndex = 0; columnIndex < columnCount; columnIndex++) {
            visitColumn(this.getColumnValues(columnIndex), columnIndex, this.getColumnVector(columnIndex).type);
        }
    }
    getColumns() {
        const columns = [];
        this.visitColumns((column) => columns.push(column));
        return columns;
    }
    convertColumns(converter) {
        const convertedColumns = [];
        const columnCount = this.columnCount;
        for (let columnIndex = 0; columnIndex < columnCount; columnIndex++) {
            convertedColumns.push(this.convertColumnValues(columnIndex, converter));
        }
        return convertedColumns;
    }
    setColumns(columns) {
        if (columns.length > 0) {
            this.rowCount = columns[0].length;
        }
        for (let columnIndex = 0; columnIndex < columns.length; columnIndex++) {
            this.setColumnValues(columnIndex, columns[columnIndex]);
        }
    }
    visitColumnMajor(visitValue) {
        const columnCount = this.columnCount;
        for (let columnIndex = 0; columnIndex < columnCount; columnIndex++) {
            this.visitColumnValues(columnIndex, visitValue);
        }
    }
    visitRowValues(rowIndex, visitValue) {
        const columnCount = this.columnCount;
        for (let columnIndex = 0; columnIndex < columnCount; columnIndex++) {
            const vector = this.getColumnVector(columnIndex);
            visitValue(vector.getItem(rowIndex), rowIndex, columnIndex, vector.type);
        }
    }
    getRowValues(rowIndex) {
        const values = [];
        this.visitRowValues(rowIndex, (value) => values.push(value));
        return values;
    }
    convertRowValues(rowIndex, converter) {
        const convertedValues = [];
        this.visitRowValues(rowIndex, (value, _, columnIndex) => convertedValues.push(converter.convertValue(value, this.getColumnVector(columnIndex).type)));
        return convertedValues;
    }
    visitRows(visitRow) {
        const rowCount = this.rowCount;
        for (let rowIndex = 0; rowIndex < rowCount; rowIndex++) {
            visitRow(this.getRowValues(rowIndex), rowIndex);
        }
    }
    getRows() {
        const rows = [];
        this.visitRows((row) => rows.push(row));
        return rows;
    }
    convertRows(converter) {
        const convertedRows = [];
        const rowCount = this.rowCount;
        for (let rowIndex = 0; rowIndex < rowCount; rowIndex++) {
            convertedRows.push(this.convertRowValues(rowIndex, converter));
        }
        return convertedRows;
    }
    setRows(rows) {
        this.rowCount = rows.length;
        const columnCount = this.columnCount;
        for (let columnIndex = 0; columnIndex < columnCount; columnIndex++) {
            const vector = this.getColumnVector(columnIndex);
            for (let rowIndex = 0; rowIndex < rows.length; rowIndex++) {
                vector.setItem(rowIndex, rows[rowIndex][columnIndex]);
            }
            vector.flush();
        }
    }
    visitRowMajor(visitValue) {
        const rowCount = this.rowCount;
        const columnCount = this.columnCount;
        for (let rowIndex = 0; rowIndex < rowCount; rowIndex++) {
            for (let columnIndex = 0; columnIndex < columnCount; columnIndex++) {
                const vector = this.getColumnVector(columnIndex);
                visitValue(vector.getItem(rowIndex), rowIndex, columnIndex, vector.type);
            }
        }
    }
}
exports.DuckDBDataChunk = DuckDBDataChunk;
