// @duckdb/node-api/lib/DuckDBConnection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DuckDBConnection = void 0;
const node_bindings_1 = __importDefault(require("@duckdb/node-bindings"));
const DuckDBAppender_1 = require("./DuckDBAppender");
const DuckDBExtractedStatements_1 = require("./DuckDBExtractedStatements");
const DuckDBMaterializedResult_1 = require("./DuckDBMaterializedResult");
const DuckDBPreparedStatement_1 = require("./DuckDBPreparedStatement");
const DuckDBResultReader_1 = require("./DuckDBResultReader");
class DuckDBConnection {
    connection;
    constructor(connection) {
        this.connection = connection;
    }
    static async create(instance) {
        return instance.connect();
    }
    /** Same as disconnect. */
    close() {
        return this.disconnect();
    }
    disconnect() {
        return node_bindings_1.default.disconnect(this.connection);
    }
    interrupt() {
        node_bindings_1.default.interrupt(this.connection);
    }
    get progress() {
        return node_bindings_1.default.query_progress(this.connection);
    }
    async run(sql, values, types) {
        if (values) {
            const prepared = await this.prepare(sql);
            prepared.bind(values, types);
            return prepared.run();
        }
        else {
            return new DuckDBMaterializedResult_1.DuckDBMaterializedResult(await node_bindings_1.default.query(this.connection, sql));
        }
    }
    async runAndRead(sql, values, types) {
        return new DuckDBResultReader_1.DuckDBResultReader(await this.run(sql, values, types));
    }
    async runAndReadAll(sql, values, types) {
        const reader = new DuckDBResultReader_1.DuckDBResultReader(await this.run(sql, values, types));
        await reader.readAll();
        return reader;
    }
    async runAndReadUntil(sql, targetRowCount, values, types) {
        const reader = new DuckDBResultReader_1.DuckDBResultReader(await this.run(sql, values, types));
        await reader.readUntil(targetRowCount);
        return reader;
    }
    async stream(sql, values, types) {
        const prepared = await this.prepare(sql);
        if (values) {
            prepared.bind(values, types);
        }
        return prepared.stream();
    }
    async streamAndRead(sql, values, types) {
        return new DuckDBResultReader_1.DuckDBResultReader(await this.stream(sql, values, types));
    }
    async streamAndReadAll(sql, values, types) {
        const reader = new DuckDBResultReader_1.DuckDBResultReader(await this.stream(sql, values, types));
        await reader.readAll();
        return reader;
    }
    async streamAndReadUntil(sql, targetRowCount, values, types) {
        const reader = new DuckDBResultReader_1.DuckDBResultReader(await this.stream(sql, values, types));
        await reader.readUntil(targetRowCount);
        return reader;
    }
    async start(sql, values, types) {
        const prepared = await this.prepare(sql);
        if (values) {
            prepared.bind(values, types);
        }
        return prepared.start();
    }
    async startStream(sql, values, types) {
        const prepared = await this.prepare(sql);
        if (values) {
            prepared.bind(values, types);
        }
        return prepared.startStream();
    }
    async prepare(sql) {
        return new DuckDBPreparedStatement_1.DuckDBPreparedStatement(await node_bindings_1.default.prepare(this.connection, sql));
    }
    async extractStatements(sql) {
        const { extracted_statements, statement_count } = await node_bindings_1.default.extract_statements(this.connection, sql);
        if (statement_count === 0) {
            throw new Error(`Failed to extract statements: ${node_bindings_1.default.extract_statements_error(extracted_statements)}`);
        }
        return new DuckDBExtractedStatements_1.DuckDBExtractedStatements(this.connection, extracted_statements, statement_count);
    }
    async createAppender(table, schema, catalog) {
        return new DuckDBAppender_1.DuckDBAppender(node_bindings_1.default.appender_create_ext(this.connection, catalog ?? null, schema ?? null, table));
    }
}
exports.DuckDBConnection = DuckDBConnection;
